> * TC_00_DispVMPerf_whonix-workstation-18: [test_413_dom0_dispvm_preload_4_api](https://openqa.qubes-os.org/tests/172098#step/TC_00_DispVMPerf_whonix-workstation-18/28) (failure)
>       `AssertionError: '/usr/lib/qubes/tests/dispvm_perf.py --dvm=test-ins...`

```
INFO:vm.disp5324:Preload startup waiting 'qubes.WaitForRunningSystem' with '60' seconds timeout
INFO:vm.disp5324:Preload startup completed 'qubes.WaitForRunningSystem'
INFO:vm.disp5324:Setting qube memory to pref mem
INFO:vm.disp1795:Preload startup waiting 'qubes.WaitForRunningSystem' with '60' seconds timeout
INFO:vm.disp1795:Preload startup completed 'qubes.WaitForRunningSystem'
```
But looking at console logs of those two disposables, I don't see why it would timeout. Startup seems to complete within 10s and there is no failures. So, maybe do extend timeout handling (in tests) with output of `systemctl` and/or `systemd-analyze` or similar?

--- a/qubes/ext/gui.py
+++ b/qubes/ext/gui.py
@@ -35,18 +35,6 @@ class GUI(qubes.ext.Extension):
             if getattr(domain, "guivm", None) and domain.guivm == vm:
                 yield domain
 
-    @qubes.ext.handler("domain-pre-shutdown")
-    def on_domain_pre_shutdown(self, vm, event, **kwargs):
-        attached_vms = [
-            domain for domain in self.attached_vms(vm) if domain.is_running()
-        ]
-        if attached_vms and not kwargs.get("force", False):
-            raise qubes.exc.QubesVMError(
-                self,
-                "There are running VMs using this VM as GuiVM: "
-                "{}".format(", ".join(vm.name for vm in attached_vms)),
-            )
-
     @staticmethod
     def send_gui_mode(vm):
         vm.run_service(
@@ -57,12 +45,6 @@ class GUI(qubes.ext.Extension):
                 else "FULLSCREEN"
             ),
         )
-
-    @qubes.ext.handler("domain-init", "domain-load")
-    def on_domain_init_load(self, vm, event):
-        if getattr(vm, "guivm", None):
-            if "guivm-" + vm.guivm.name not in vm.tags:
-                self.on_property_set(vm, event, name="guivm", newvalue=vm.guivm)
 
     @qubes.ext.handler("property-reset:guivm")
     def on_property_reset(self, subject, event, name, oldvalue=None):
@@ -81,6 +63,51 @@ class GUI(qubes.ext.Extension):
         if newvalue:
             guivm = "guivm-" + newvalue.name
             subject.tags.add(guivm)
+
+    @qubes.ext.handler("property-set:default_guivm", system=True)
+    def on_property_set_default_guivm(
+        self, app, event, name, newvalue, oldvalue=None
+    ):
+        for vm in app.domains:
+            if hasattr(vm, "guivm") and vm.property_is_default("guivm"):
+                vm.fire_event(
+                    "property-set:guivm",
+                    name="guivm",
+                    newvalue=newvalue,
+                    oldvalue=oldvalue,
+                )
+
+    @qubes.ext.handler("property-reset:keyboard_layout")
+    def on_keyboard_reset(self, vm, event, name, oldvalue=None):
+        if not vm.is_running():
+            return
+        kbd_layout = vm.keyboard_layout
+
+        vm.untrusted_qdb.write("/keyboard-layout", kbd_layout)
+
+    @qubes.ext.handler("property-set:keyboard_layout")
+    def on_keyboard_set(self, vm, event, name, newvalue, oldvalue=None):
+        if newvalue == oldvalue:
+            return
+
+        if vm.is_running():
+            vm.untrusted_qdb.write("/keyboard-layout", newvalue)
+
+        for domain in vm.app.domains:
+            if getattr(
+                domain, "guivm", None
+            ) == vm and domain.property_is_default("keyboard_layout"):
+                domain.fire_event(
+                    "property-reset:keyboard_layout",
+                    name="keyboard_layout",
+                    oldvalue=oldvalue,
+                )
+
+    @qubes.ext.handler("domain-init", "domain-load")
+    def on_domain_init_load(self, vm, event):
+        if getattr(vm, "guivm", None):
+            if "guivm-" + vm.guivm.name not in vm.tags:
+                self.on_property_set(vm, event, name="guivm", newvalue=vm.guivm)
 
     @qubes.ext.handler("domain-qdb-create")
     def on_domain_qdb_create(self, vm, event):
@@ -118,18 +145,16 @@ class GUI(qubes.ext.Extension):
                 "/guivm-windows-prefix", guivm_windows_prefix
             )
 
-    @qubes.ext.handler("property-set:default_guivm", system=True)
-    def on_property_set_default_guivm(
-        self, app, event, name, newvalue, oldvalue=None
-    ):
-        for vm in app.domains:
-            if hasattr(vm, "guivm") and vm.property_is_default("guivm"):
-                vm.fire_event(
-                    "property-set:guivm",
-                    name="guivm",
-                    newvalue=newvalue,
-                    oldvalue=oldvalue,
-                )
+    @qubes.ext.handler("domain-tag-add:created-by-*")
+    def set_guivm_on_created_by(self, vm, event, tag, **kwargs):
+        """Set GuiVM based on 'tag-created-vm-with' and 'set-created-guivm'
+        features
+        """
+        # pylint: disable=unused-argument
+        created_by = vm.app.domains[tag.partition("created-by-")[2]]
+        if created_by.features.get("set-created-guivm", None):
+            guivm = vm.app.domains[created_by.features["set-created-guivm"]]
+            vm.guivm = guivm
 
     @qubes.ext.handler("domain-start")
     async def on_domain_start(self, vm, event, **kwargs):
@@ -149,39 +174,14 @@ class GUI(qubes.ext.Extension):
                 "/usr/bin/qubes-input-trigger", "--all", "--dom0"
             )
 
-    @qubes.ext.handler("property-reset:keyboard_layout")
-    def on_keyboard_reset(self, vm, event, name, oldvalue=None):
-        if not vm.is_running():
-            return
-        kbd_layout = vm.keyboard_layout
-
-        vm.untrusted_qdb.write("/keyboard-layout", kbd_layout)
-
-    @qubes.ext.handler("property-set:keyboard_layout")
-    def on_keyboard_set(self, vm, event, name, newvalue, oldvalue=None):
-        if newvalue == oldvalue:
-            return
-
-        if vm.is_running():
-            vm.untrusted_qdb.write("/keyboard-layout", newvalue)
-
-        for domain in vm.app.domains:
-            if getattr(
-                domain, "guivm", None
-            ) == vm and domain.property_is_default("keyboard_layout"):
-                domain.fire_event(
-                    "property-reset:keyboard_layout",
-                    name="keyboard_layout",
-                    oldvalue=oldvalue,
-                )
-
-    @qubes.ext.handler("domain-tag-add:created-by-*")
-    def set_guivm_on_created_by(self, vm, event, tag, **kwargs):
-        """Set GuiVM based on 'tag-created-vm-with' and 'set-created-guivm'
-        features
-        """
-        # pylint: disable=unused-argument
-        created_by = vm.app.domains[tag.partition("created-by-")[2]]
-        if created_by.features.get("set-created-guivm", None):
-            guivm = vm.app.domains[created_by.features["set-created-guivm"]]
-            vm.guivm = guivm
+    @qubes.ext.handler("domain-pre-shutdown")
+    def on_domain_pre_shutdown(self, vm, event, **kwargs):
+        attached_vms = [
+            domain for domain in self.attached_vms(vm) if domain.is_running()
+        ]
+        if attached_vms and not kwargs.get("force", False):
+            raise qubes.exc.QubesVMError(
+                self,
+                "There are running VMs using this VM as GuiVM: "
+                "{}".format(", ".join(vm.name for vm in attached_vms)),
+            )

--- a/qubes/ext/gui.py
+++ b/qubes/ext/gui.py
@@ -32,7 +32,7 @@ class GUI(qubes.ext.Extension):
     @staticmethod
     def attached_vms(vm):
         for domain in vm.app.domains:
-            if getattr(domain, "guivm", None) and domain.guivm == vm:
+            if hasattr(domain, "guivm") and domain.guivm is vm:
                 yield domain
 
     @staticmethod
@@ -56,10 +56,15 @@ class GUI(qubes.ext.Extension):
         # Clean other 'guivm-XXX' tags.
         # gui-daemon can connect to only one domain
         tags_list = list(subject.tags)
+        found = False
         for tag in tags_list:
             if tag.startswith("guivm-"):
+                if newvalue and tag == "guivm-" + newvalue.name:
+                    found = True
+                    continue
                 subject.tags.remove(tag)
-
+        if found:
+            return
         if newvalue:
             guivm = "guivm-" + newvalue.name
             subject.tags.add(guivm)
@@ -82,32 +87,33 @@ class GUI(qubes.ext.Extension):
         if not vm.is_running():
             return
         kbd_layout = vm.keyboard_layout
-
         vm.untrusted_qdb.write("/keyboard-layout", kbd_layout)
 
     @qubes.ext.handler("property-set:keyboard_layout")
     def on_keyboard_set(self, vm, event, name, newvalue, oldvalue=None):
         if newvalue == oldvalue:
             return
-
         if vm.is_running():
             vm.untrusted_qdb.write("/keyboard-layout", newvalue)
-
-        for domain in vm.app.domains:
-            if getattr(
-                domain, "guivm", None
-            ) == vm and domain.property_is_default("keyboard_layout"):
-                domain.fire_event(
-                    "property-reset:keyboard_layout",
-                    name="keyboard_layout",
-                    oldvalue=oldvalue,
-                )
+        attached_vms = [
+            domain
+            for domain in self.attached_vms(vm)
+            if domain.property_is_default("keyboard_layout")
+        ]
+        for domain in attached_vms:
+            domain.fire_event(
+                "property-reset:keyboard_layout",
+                name="keyboard_layout",
+                oldvalue=oldvalue,
+            )
 
     @qubes.ext.handler("domain-init", "domain-load")
     def on_domain_init_load(self, vm, event):
-        if getattr(vm, "guivm", None):
-            if "guivm-" + vm.guivm.name not in vm.tags:
-                self.on_property_set(vm, event, name="guivm", newvalue=vm.guivm)
+        guivm = getattr(vm, "guivm", None)
+        if not guivm:
+            return
+        if "guivm-" + guivm.name not in vm.tags:
+            self.on_property_set(vm, event, name="guivm", newvalue=guivm)
 
     @qubes.ext.handler("domain-qdb-create")
     def on_domain_qdb_create(self, vm, event):
@@ -120,27 +126,23 @@ class GUI(qubes.ext.Extension):
             except KeyError:
                 pass
 
-        vm.untrusted_qdb.write(
-            "/qubes-gui-enabled",
-            str(
-                bool(
-                    getattr(vm, "guivm", None) and vm.features.get("gui", True)
-                )
-            ),
-        )
+        guivm = getattr(vm, "guivm", None)
+        gui = bool(guivm and vm.features.get("gui", True))
+        vm.untrusted_qdb.write("/qubes-gui-enabled", str(gui))
         # Add GuiVM Xen ID for gui-daemon
-        if getattr(vm, "guivm", None):
+        if guivm:
             if vm != vm.guivm:
                 vm.untrusted_qdb.write("/keyboard-layout", vm.keyboard_layout)
-
                 if vm.guivm.is_running():
                     vm.untrusted_qdb.write(
                         "/qubes-gui-domain-xid", str(vm.guivm.xid)
                     )
 
         # Set GuiVM prefix
-        guivm_windows_prefix = vm.features.get("guivm-windows-prefix", "GuiVM")
         if vm.features.get("service.guivm", None):
+            guivm_windows_prefix = vm.features.get(
+                "guivm-windows-prefix", "GuiVM"
+            )
             vm.untrusted_qdb.write(
                 "/guivm-windows-prefix", guivm_windows_prefix
             )
@@ -152,9 +154,8 @@ class GUI(qubes.ext.Extension):
         """
         # pylint: disable=unused-argument
         created_by = vm.app.domains[tag.partition("created-by-")[2]]
-        if created_by.features.get("set-created-guivm", None):
-            guivm = vm.app.domains[created_by.features["set-created-guivm"]]
-            vm.guivm = guivm
+        if created_guivm := created_by.features.get("set-created-guivm", None):
+            vm.guivm = vm.app.domains[created_guivm]
 
     @qubes.ext.handler("domain-start")
     async def on_domain_start(self, vm, event, **kwargs):
@@ -162,10 +163,8 @@ class GUI(qubes.ext.Extension):
             domain for domain in self.attached_vms(vm) if domain.is_running()
         ]
         for attached_vm in attached_vms:
-            attached_vm.untrusted_qdb.write(
-                "/qubes-gui-enabled",
-                str(bool(attached_vm.features.get("gui", True))),
-            )
+            gui = bool(attached_vm.features.get("gui", True))
+            attached_vm.untrusted_qdb.write("/qubes-gui-enabled", str(gui))
             attached_vm.untrusted_qdb.write(
                 "/qubes-gui-domain-xid", str(vm.xid)
             )

--- a/qubes/vm/dispvm.py
+++ b/qubes/vm/dispvm.py
@@ -506,8 +506,7 @@ class DispVM(qubes.vm.qubesvm.QubesVM):
             await asyncio.wait_for(
                 self.run_service_for_stdio(
                     service,
-                    stdout=subprocess.DEVNULL,
-                    stderr=subprocess.DEVNULL,
+                    stderr=subprocess.STDOUT,
                 ),
                 timeout=timeout,
             )
@@ -519,12 +518,13 @@ class DispVM(qubes.vm.qubesvm.QubesVM):
                 "startup. To debug, run the following on a new disposable of "
                 "'%s': %s" % (service, timeout, self.template, debug_msg)
             )
-        except (subprocess.CalledProcessError, qubes.exc.QubesException):
-            debug_msg = "systemctl --failed"
+        except subprocess.CalledProcessError as e:
             raise qubes.exc.QubesException(
-                "Error on call to '%s' during preload startup. To debug, "
-                "disable preloading from '%s' and run the following on a new "
-                "disposable: %s" % (service, self.template, debug_msg)
+                "Error on call to '%s' during preload startup: %s"
+                % (
+                    service,
+                    qubes.utils.sanitize_stderr_for_log(e.stdout),
+                )
             )
 
     @qubes.events.handler("domain-start")

